pkg/utils/k8sutil.go

```diff
@@ -937,9 +1080,9 @@ func CreateServiceMonitor(smclient monitoringv1alpha1.MonitoringV1alpha1Client,
 	return fmt.Errorf("service monitor has already existed")
 }
 
-// GetOwnerReference returns ownerRef for appending to objects's metadata
-// created by kubeless-controller one a function is deployed.
-func GetOwnerReference(funcObj *kubelessApi.Function) ([]metav1.OwnerReference, error) {
+// GetFunctionOwnerReference returns ownerRef for appending to objects's metadata
+// created by kubeless-controller once a function is deployed.
+func GetFunctionOwnerReference(funcObj *kubelessApi.Function) ([]metav1.OwnerReference, error) {
```

Comment:
This can be made generic across the types using something like this (uncompiled, so probably contains typos):

```go
import "k8s.io/apimachinery/pkg/api/meta"

func GetOwnerReference(obj runtime.Object) ([]metav1.OwnerReference, error) {
   meta, err := meta.Accessor(obj)
   if err != nil {
      return nil, err
   }
   if meta.GetName() == "" {
      // return name can't be empty
   }
   if meta.GetUID() == "" {
      // return uid can't be empty
   }
   or := metav1.NewControllerRef(meta, obj.GetObjectKind().GroupVersionKind())
   return []metav1.OwnerReference{*or}
}
```

Note also that this returns a *controller* ref, and sets `BlockOwnerDeletion` and `Controller` to `true`.  I think this is desirable for server-side garbage collection, but if we don't want this, then it's still easy to use the generic accessors for the other OwnerReference fields.

Reply:
This makes sense. I tried to do it this way. But created `OwnerReference` did not have `api version` etc not set. I tried to use `meta.TypeAccessor` to get the api version. But could not get something wokring consistenly. Could not find the root cause. I will try to revisit this.